### PR TITLE
tenantcapabilities: introduce a Watcher over system.tenants

### PR DESCRIPTION
**Pull request by arulajmani**
This patch introduces three new interfaces -- a Watcher, a Reader,
and Authorizer. They're not hooked up yet, but once they are, they'll
work together to provide (in-memory) capability checks for incoming
tenant requests.

The Watcher establishes a rangefeed over `system.tenants` to
incrementally (and transparently) maintain an in-memory view of the
global tenant capability state. Publicly, it exposes a `Reader`
interface.

The `Reader` provides access to the global tenant capability state.
The `Watcher` and `Authorizer` communicate with each other using the
`Reader` interface.

The `Authorizer` consulsts the global tenant capability state to perform
authorization checks for incoming requests issued by tenants. Part of
the motivation to structure the code as such is to expand the set of
inputs the `Authorizer` uses to authorize requests. One could imagine
other dependencies being injected into the `Authorizer` in the future.

Epic: CRDB-18503
References: https://github.com/cockroachdb/cockroach/issues/94643

Release note: None

### CODE REVIEW

**Review by knz — COMMENTED** [review]
I'm OK with the overall architecture of the watcher, and with the approach to testing.
Here are my major comments so far.

1. I'm not sure we can reasonably expect the Authorizer to be implemented in a self-contained package over time. IMHO it will need to be part of either pkg/server or pkg/kv/kvserver, because it will grow over time to include parameters from the environment (things like resource usage, pattern of previous requests etc). I have no objections to placing the _interface_ in a separate package but putting the implementation there is dubious (to me)

2. it's going to be very important to have more members of the KV team look at this code and participate in the review iteration, because the KV team as a whole will bear the responsibility to maintain this forward.

I'd like to iterate on point (1) before this PR can be merged. Point (2) may not be a blocker to the merge, but be wary of adding much KV code without creating a sense of ownership to the rest of the KV team; otherwise this could come back to bite us later.

> Reviewed 18 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NLWg3CQES85UejyNdvm:bta52yn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @ecwall)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLWdtFQ0wCG1FmKI1GB:-NLWdtFQ0wCG1FmKI1GC:be20vm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L102)):*
> ```Go
> 	deleted := !ev.Value.IsPresent()
> 	var value roachpb.Value
> 	if deleted {
> ```

I don't understand what this condition is doing. This function could use a few more explanatory comments.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLWdy6q3bluXu0Hh_3q:-NLWdy6r2urW1FWDnfwW:bxgefqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L116)):*
> ```Go
> 		Value: value,
> 	})
> ```

nit: remove empty line.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLWfYxxCnuGXwpVDvyi:-NLWfYxy9Xkz02hQFQ1Y:b-qa4s61) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L53)):*
> ```Go
> 	bufferMemLimit int64,
> 	knobs *tenantcapabilities.TestingKnobs,
> ) *Watcher {
> ```

you can return the interface type here to hide implementation details.

**Review by arulajmani — COMMENTED** [review]
We can talk about point (1) a bit more in-person, during tomorrow's pod meeting. I do want to explore how we could keep the `Authorizer` implementation self-contained, if at all possible. Could we instead inject interfaces which give the `Authorizer` access to whatever parameters it needs from the environment when we get there? Maybe I'm oversimplifying this because I don't appreciate how some of the capabilities we might add in the future might look in code.  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NLXdvd7E14h2g5h749t:b-m8v4ns)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall)

**Comment by knz** [comment]
> Could we instead inject interfaces which give the Authorizer access to whatever parameters it needs from the environment when we get there? 

yes that sounds reasonable. Do you have an idea (if only a sketch) of how it would look like?

**Review by arulajmani — COMMENTED** [review]
> Do you have an idea (if only a sketch) of how it would look like?

I think instead of having the `Watcher` call into `tenantcapabilitiesauthorizer.New`, it instead needs to be passed in the `Authorizer` [1] or an `AuthorizerFactory`. Then, whatever dependencies we need for the `Authorizer`, would be passed to it during creation (or in the creation of the `AuthorizerFactory`) .

[1] If we go this route, we would need to add a `Reset` method on the `Authorizer`, to have it throw away its state. We need this if, for some reason, the rangefeed runs into an error and as a result performs a full scan of `system.tenants`. 

I have a mild preference for the `AuthorizerFactory` approach. But I could also go the other way, given it feels a bit premature for now, without any dependencies to speak of. What do you think? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NLXuvfX4LrictMrpk3v:bydcumz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @knz)

**Review by knz — COMMENTED** [review]
My expectation based on other subsystems is that the Reset route is better -- the authorizer will have other state besides whatever cache it will maintain, and re-instantiating it on every invalidation will be expensive.

But I agree it could go either way.

At this point I think the most important to decouple the two and make the authorizer/watcher connection work via an interface, instead of a direct package dependency.

Btw my other comments from before still apply.

> Reviewed 1 of 19 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NLfH750CKWSmUdRrjWq:b-gcwq2a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @ecwall)

**Review by ecwall — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLg4Xtp84QhXBnr1HJT:-NLg4Xtq4MxS6VZnGVX-:b-ut4544) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/capabilities.go#L24)):*
> ```Go
> // by watching for changes.
> type CapabilitiesWatcher interface {
> 	Authorizer
> ```

Why is the watcher also an authorizer?

It looks like `CapabilitiesWatcher` is the write interface and `Authorizer` is the read interface?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLg50MqDFVBYndSubpP:-NLg50MqDFVBYndSubpQ:buew706) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L55)):*
> ```Go
> 		switch ru.GetInner().(type) {
> 		case *roachpb.AdminSplitRequest:
> 			cp, ok := a.mu.capabilities[tenID]
> ```

The `RLock()` only needs to guard this line and it can happen once before the loop.

`cp` will be a copy of what is in the map so it does not need to be guarded.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLg3J2HDvvBFdwxS8bo:-NLg3J2HDvvBFdwxS8bp:bb7phvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L37)):*
> ```Go
> 	bufferMemLimit   int64
> 
> 	tenantsTableID uint32 // overriden for tests
> ```

typo

**Review by arulajmani — COMMENTED** [review]
After mulling over this a bit more, and keeping in mind we want the `Authorizer` to be general enough that it could be extended to use other signals to perform authorization in the future, I decided to add a new `Reader` interface. We then use this `Reader` interface to communicate between the `Authorizer` and `Watcher` -- the `Watcher` is responsible for incrementally maintaining the underlying tenant capabilities state, and exposes it via the `Reader`. Conversely, the `Authorizer` consults the global tenant capability state to make authorization decisions. Let me know what you think! 

> Dismissed @ecwall from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NM7FZ9O34Wh__WDkmZT:bd9jprt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @knz)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLg4Xtp84QhXBnr1HJT:-NM7Ddvl-BZLKcipkrA4:bu6ky4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/capabilities.go#L24)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Why is the watcher also an authorizer?

It looks like `CapabilitiesWatcher` is the write interface and `Authorizer` is the read interface?
</blockquote></details>

Changed some of this stuff around, so this comment no longer applies. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLg50MqDFVBYndSubpP:-NM7DkGm-EB_w6w8TMf8:bco6wfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L55)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

The `RLock()` only needs to guard this line and it can happen once before the loop.

`cp` will be a copy of what is in the map so it does not need to be guarded.
</blockquote></details>

Changed some of this stuff around, but the new changes follow the ethos of your comment. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go` line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLWdtFQ0wCG1FmKI1GB:-NM-N-H01z3I3jxnt1NG:b-qy7sib) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/decoder.go#L102)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I don't understand what this condition is doing. This function could use a few more explanatory comments.
</blockquote></details>

Added. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLg3J2HDvvBFdwxS8bo:-NM-NB7J0IC6-8nwUpgm:bjrf1k4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L37)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Which part?
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go` line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NLWfYxxCnuGXwpVDvyi:-NM7I8c1DlJHrQ9fSpWD:b82655l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9772b3f37a909d3f30a97a0204d2cca208e78100/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher.go#L53)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

you can return the interface type here to hide implementation details.
</blockquote></details>

I was intending to accept the value returned of this function as the interface, instead of returning an interface here. I don't think the whole "return interface vs. structs" matters much for what we're doing here, but the one nicety of returning the concrete struct here is it allows us to access the testing function directly, without having to perform a cast or pollute the interface.

**Review by knz — APPROVED** [review]
:lgtm: 
Before you merge this, can you do a back-of-the-envelope estimation of how much memory this uses for, say, 10000 tenant records and 100 bytes of capability payload per tenant - and mention it in the PR description.

> Reviewed 15 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NM8F8df-0qh1c_UUnPg:bmo9unn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @irfansharif)

___
*[`-- commits` line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NM8ESJY7mt9ITOtvry3):*
consults*

**Review by arulajmani — COMMENTED** [review]
Added some words. I chose to use 50 bytes per capability payload, given that's the constant we use in the PR to allocate the buffer. 

TFTR! 

bors r=knz 

> Dismissed @knz from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-:-NMAMU0J5IcjN87JD2oY:b-hpan0c)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ecwall and @irfansharif)

___
*[`-- commits` line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95040#-NM8ESJY7mt9ITOtvry3):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

consults*
</blockquote></details>

Done.

**Comment by arulajmani** [comment]
bors r=knz

**Review by ecwall — APPROVED** [review]
> Reviewed 6 of 19 files at r1, 15 of 15 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95040)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif)

File: pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Test basic tenant capability watcher functionality.
+
+start
+----
+ok
+
+updates
+----
+
+update-state
```
Comment (irfansharif) [inline pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/basic:10]: Why is there no "Complete Update" here? It's completely updated, right? To an empty table.

**Review by irfansharif — APPROVED** [review]
This code smells vaguely familiar 😛

**Pull request merged**